### PR TITLE
Port Measured rootfs feature from CCv0 branch to main

### PR DESCRIPTION
**Pull request by arronwy**
Current this feature is need build with `KATA_BUILD_CC`, we change to other flag if needed.
Verified with below commands: 
```
export KATA_BUILD_CC=yes
./tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh --build=kernel
./tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh --build=rootfs-image
./tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh --build=shim-v2
```

Fixes: #6674

### CODE REVIEW

**Comment by wainersm** [comment]
@fidencio we don't have any pipeline that builds with `KATA_BUILD_CC=yes`, do we? Asking because it would be nice to have at least one to test the initramfs build.

Ah, another thing... when we first thought about merging the branches, we thought about having CoCo specific builds of kata containers, that's why I introduced the `KATA_BUILD_CC` variable on the main branch. However, I don't know if things changed...what do you plan on that? A single Kata build? If so, then probably @arronwy should be using a different variable (`MEASURED_BOOT`?) to turn on/off this feature.

Cc @stevenhorsman

**Comment by fitzthum** [comment]
Just mentioning in case it is useful in discussion, but this feature is not compatible with SEV or SNP and most likely won't be in the future either. Nothing against this PR, but that might be good info for people to be aware of.

File: tools/packaging/static-build/shim-v2/build.sh [hunk 1]
```diff
@@ -21,10 +21,17 @@ DESTDIR=${DESTDIR:-${PWD}}
 PREFIX=${PREFIX:-/opt/kata}
 container_image="${SHIM_V2_CONTAINER_BUILDER:-$(get_shim_v2_image_name)}"
 
+EXTRA_OPTS="${EXTRA_OPTS:-""}"
+
+http_proxy="${http_proxy:-}"
+https_proxy="${https_proxy:-}"
```
Comment (mythi) [inline tools/packaging/static-build/shim-v2/build.sh:27]: better keep it consistent with other `build.sh` scripts and leave the proxy settings out?

**Comment by fidencio** [comment]
> > Just mentioning in case it is useful in discussion, but this feature is not compatible with SEV or SNP and most likely won't be in the future either. Nothing against this PR, but that might be good info for people to be aware of.
> 
> @fitzthum is it just the way it's implemented here or dm-verity cannot be used for some reason?

@fitzthum ? :-)

**Comment by fitzthum** [comment]
> @fitzthum is it just the way it's implemented here or dm-verity cannot be used for some reason?

Whoops, didn't see this. 

So by default when using SEV and SNP with QEMU the hardware only measures the initial flash0 memory of the guest, which is just the firmware. To extend the TCB we use a so-called "measured direct boot" approach, where QEMU injects the hashes the the initrd, kernel, and command line, into the firmware. Since the firmware is measured, these hashes are measured. Then on boot the firmware makes sure that the hashes match the components supplied via firmware config interface. This is why we use a separate (upstream) build of OVMF, btw.

When we use measured direct boot, we already measure the initrd, so the dm-verity approach seems redundant. If we don't use measured direct boot, the kernel parameters aren't measured by the hw, so dm-verity wouldn't be tied to the root of trust.

It is possible that we could change the measured direct boot implementation to measure the kernel command line and kernel but not the initrd and then use dm-verity for that, but I don't see any reason to do this. 

With SNP we will use the same measured direct boot approach or a TPM-based approach. I think the TPM-based approach will have other means for measuring the initrd.

**Comment by wainersm** [comment]
> @arronwy, I've made a few comments, and I'd like to ask you for a few changes here as well.
> 
>     1. Please, add a `tdx-rootfs-image-tarball` which can take advantage of this
> 

The rootfs for sev will be named rootfs-initrd-sev (so `rootfs-initrd-sev-tarball` target). Could we follow that naming pattern for TDX too?

>     2. Make shim-v2-tarball depend on `tdx-rootfs-image-tarball` and `rootfs-image-tarball`
>

Could you please expand on that idea @fidencio ? This will make a core (supposed to be) TEE-agnostic component to depend on the TDX build path.
  
> 
> We'll also need changes in the caching mechanism, but those ones I can help with, as I already did with the `nvidia-gpu` artefacts.
> 
> I saw that both Wainer and Steven raised the "testing" point, which is super valid. I think this can be easily tested with `ctr`, and we can add a test case for that in the tests repo (for now, then I will migrate that to GitHub actions as time allows). The test doesn't need to be done / run for TDX, as this is a general feature rather than a TEE one.
> 

Yes, tests, please :)

I would advocate for documenting this feature as well.

> @wainersm, @stevenhorsman, @arronwy, WDYT?

File: tools/packaging/static-build/shim-v2/build.sh (hunk 1 above)
Comment (arronwy) [reply tools/packaging/static-build/shim-v2/build.sh]: Thanks @mythi , fixed now.

File: tools/osbuilder/image-builder/image_builder.sh [hunk 2]
```diff
@@ -11,6 +11,7 @@ set -o errexit
 set -o pipefail
 
 DOCKER_RUNTIME=${DOCKER_RUNTIME:-runc}
+KATA_BUILD_CC=${KATA_BUILD_CC:-no}
```
[the remark this answers is not in the document]
Comment (arronwy) [reply tools/osbuilder/image-builder/image_builder.sh]: Thanks @fidencio  @wainersm , switched to `MEASURED_ROOTFS` now :)

File: tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh [hunk 3]
```diff
@@ -407,7 +414,20 @@ install_shimv2() {
 
 	export GO_VERSION
 	export RUST_VERSION
-	DESTDIR="${destdir}" PREFIX="${prefix}" "${shimv2_builder}"
+
+	if [ "${KATA_BUILD_CC}" == "yes" ]; then
+	        extra_opts="DEFSERVICEOFFLOAD=true"
+		if [ -f "${repo_root_dir}/tools/osbuilder/root_hash.txt" ]; then
+			root_hash=$(sudo sed -e 's/Root hash:\s*//g;t;d' "${repo_root_dir}/tools/osbuilder//root_hash.txt")
+			root_measure_config="cc_rootfs_verity.scheme=dm-verity cc_rootfs_verity.hash=${root_hash}"
```
[the remark this answers is not in the document]
Comment (arronwy) [reply tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh]: fixed now :)

**Comment by arronwy** [comment]
Thanks @fidencio @wainersm @stevenhorsman @mythi @fitzthum for reviewing:
Below are main changes:
1. rebase to latest main branch
2. sed -i 's/KATA_BUILD_CC/MEASURED_ROOTFS/g' patches/*
3. sed -i 's/cc_rootfs_verity/rootfs_verity/g' patches/*
4. Add new `rootfs-image-tdx` target

export KATA_BUILD_CC=yes
./tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh --build=kernel
./tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh --build=rootfs-image-tdx
./tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh --build=shim-v2

./opt/kata/share/kata-containers/kata-containers-tdx.img
./opt/kata/share/kata-containers/kata-ubuntu-latest-tdx.image

we can also
cd tools/packaging/kata-deploy/local-build
make rootfs-image-tdx-tarball

We can submit tests and documents in another PR.

**Comment by fidencio** [comment]
/test

I've rebased Arron's commit atop of `main` as it had a conflict.

**Review by fidencio — APPROVED** [review]
lgtm, thanks @arronwy!

**Comment by fidencio** [comment]
Firecracker failure is also related to the devmapper breakage.

**Comment by bergwolf** [comment]
@arronwy @fitzthum Not to block the PR, but can we have a document to explain the different approach taken by tdx and sev, and how users can config which to use when building the rootfs image?

**Review by bergwolf — APPROVED** [review]

**Pull request merged**